### PR TITLE
This Github Action automatically pushes image into Quay.io registry

### PR DESCRIPTION
**Pull request by phracek**
Actions which are used are from GitHub Marketplace
https://github.com/marketplace/actions/buildah-build
https://github.com/marketplace/actions/push-to-registry

Signed-off-by: Petr "Stone" Hracek <phracek@redhat.com>

### CODE REVIEW

**Comment by phracek** [comment]
Please look at this PR which introduces automatic image pushing into Quay.io.

Propose a better BASH way if it is feasible.

File: .github/workflows/build-and-push.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+          versions_row=${versions_row/VERSIONS = /}
+          versions_row=$(echo $versions_row | tr " " "\n")
+          VERSIONS_MATRIX=""
+          for row in $versions_row; do
+            if [[ $row =~ "minimal" ]]; then
+              continue
+            fi
+            VERSIONS_MATRIX="$VERSIONS_MATRIX$row,"
+          done
+          # Remove last dash from string which now looks like X,Y,Z,
```
Comment (frenzymadness) [inline .github/workflows/build-and-push.yml:32]:
```suggestion
          # Remove last comma from string which now looks like X,Y,Z,
```

**Review by frenzymadness — COMMENTED** [review]
I haven't found anything wrong. Given the fact that this is only for Centos 7, will there be also an action for other containers? I mean, the name of the action does not tell that it's only for Centos 7.

**Comment by phracek** [comment]
> I haven't found anything wrong. Given the fact that this is only for Centos 7, will there be also an action for other containers? I mean, the name of the action does not tell that it's only for Centos 7.

Good point. Automatic pushing into RHEL is done by internal tools. Pushing into Fedora is not possible, because of karma etc.

All other containers in https://github.com/sclorg or will use the same GitHub action. But s2i-python-container is PoC.

File: .github/workflows/build-and-push.yml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      image_name: ${{ steps.setup-job.outputs.image_name }}
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Get all matrix versions and base image name
+        id: setup-job
+        run: |
+          # This command returns row with VERSIONS
+          versions_row=$(grep "VERSIONS" Makefile)
+          # Return only version string, spaces are separated by dash
```
[the remark this answers is not in the document]
Comment (phracek) [reply .github/workflows/build-and-push.yml:22]: I will rewrite it a bit more.

[the remark this answers is not in the document]
Comment (phracek) [reply .github/workflows/build-and-push.yml]: Rewritten a bit more.

**Comment by pkubatrh** [comment]
How about the new upstream container images, like python-39-minimal? Will they be covered by the same action?

**Comment by phracek** [comment]
@pkubatrh Thanks for this question. There will be definitely another GitHub Action. I will do it as soon as this is gonna merged. Because of quay.io registry will be different. Or this can be built only with a condition https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions.

I would prefer to do it as a separate PR so we can see that this is working properly.

**Comment by phracek** [comment]
Some example is here: https://www.edwardthomson.com/blog/github_actions_13_conditionals.html

**Comment by phracek** [comment]
The pull request is ready for merge. redis-container already uses the same GitHub Action https://github.com/sclorg/redis-container/runs/3621309385 and https://github.com/sclorg/redis-container/runs/3621309480

**Comment by phracek** [comment]
@frenzymadness The Github action for build and push is ready for merge. What do you think?

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
[test]

File: .github/workflows/build-and-push.yml [hunk 3]
```diff
@@ -22,12 +24,23 @@ jobs:
           BASE_IMAGE_ROW=${row/BASE_IMAGE_NAME = /}
           echo ::set-output name=image_name::$BASE_IMAGE_ROW
 
+      - name: Checkout into generated section
+        run: |
+          git checkout --progress --force -B generated refs/remotes/origin/generated
```
Comment (frenzymadness) [inline .github/workflows/build-and-push.yml:29]: Do we really want to use the generated branch? I mean, I don't use it and I'm not sure there is a check that the branch is always up to date.

Comment (phracek) [reply .github/workflows/build-and-push.yml:29]: It is a good question. I did not test whether dist-gen works on Ubuntu, yet ;) I can try it and then we can call 'make generate-all' and take the resources correspondingly. But it needs time to investigate it.

Comment (phracek) [reply .github/workflows/build-and-push.yml]: Fix by the latest commit. We generate sources by each merge.

**Comment by phracek** [comment]
> I've posted just one question. Otherwise, I'm not able to check it so we can merge it and see whether it works or not. Also, if we merge it before the minimal container, should I add it to the matrix?

The minimal container will have alone build-push action.

**Comment by phracek** [comment]
[test]

**Comment by frenzymadness** [comment]
We should sync about this. I don't think that generating sources via distgen is necessary but I might be missing something.

**Pull request merged**